testTop/pyTestsApp/gateway_tests/conftest.py

```diff
@@ -0,0 +1,993 @@
+        pvlist=config.default_pvlist,
+        db_file=config.test_ioc_db,
+    ) as env:
+        yield env
+
+
+@contextlib.contextmanager
+def custom_environment(
+    access_contents: str,
+    pvlist_contents: str,
+    db_contents: str = "",
+    db_file: Optional[str] = config.test_ioc_db,
+    dbd_file: Optional[str] = None,
+    encoding: str = "latin-1",
+    ioc_args: Optional[List[str]] = None,
+    gateway_args: Optional[List[str]] = None,
+) -> Generator[EnvironmentInfo, None, None]:
+    """
+    Run a gateway and an IOC in a custom environment, specifying the raw
+    contents of the access control file and the pvlist.
+
+    Parameters
+    ----------
+    access_contents : str, optional
+        The gateway access control configuration contents.
+
+    pvlist_contents : str, optional
+        The gateway pvlist configuration contents.
+
+    db_contents : str, optional
+        Additional database text to add to ``db_file``, if specified.
+
+    db_file : str, optional
+        Path to the IOC database.  Defaults to ``test_ioc_db``.  This is loaded
+        in addition to ``db_contents``, if specified.
+
+    dbd_file : str, optional
+        Path to the IOC database definition.  Defaults to using the database
+        definition provided with epics-base/softIoc.
+    """
+    gateway_args = gateway_args or []
+    ioc_args = ioc_args or []
+    if db_file is not None:
+        with open(db_file, "rt") as fp:
+            existing_db_contents = fp.read()
+        db_contents = "\n".join((existing_db_contents, textwrap.dedent(db_contents)))
+
+    access_contents = textwrap.dedent(access_contents)
+    pvlist_contents = textwrap.dedent(pvlist_contents)
+    db_contents = textwrap.dedent(db_contents)
+    with (
+        tempfile.NamedTemporaryFile() as access_fp,
+        tempfile.NamedTemporaryFile() as pvlist_fp,
+        tempfile.NamedTemporaryFile() as dbfile_fp,
+    ):
+
+        access_fp.write(access_contents.encode(encoding))
+        access_fp.flush()
+
+        pvlist_fp.write(pvlist_contents.encode(encoding))
+        pvlist_fp.flush()
+
+        dbfile_fp.write(db_contents.encode(encoding))
+        dbfile_fp.flush()
+
+        logger.info(
+            "Access rights:\n%s",
+            textwrap.indent(access_contents, "    "),
+        )
+        logger.info(
+            "PVList:\n%s",
+            textwrap.indent(pvlist_contents, "    "),
+        )
+        with (
+            run_gateway(*gateway_args, access=access_fp.name, pvlist=pvlist_fp.name),
+            run_ioc(*ioc_args, db_file=dbfile_fp.name, dbd_file=dbd_file),
+            local_channel_access(),
+        ):
+            yield EnvironmentInfo(
+                access=access_fp.name,
+                pvlist=pvlist_fp.name,
+                db_file=dbfile_fp.name,
+                dbd_file=dbd_file,
+            )
+
+
+class PyepicsCallback(Protocol):
+    def __call__(self, pvname: str = "", value: Any = None, **kwargs) -> None:
+        ...
```

Comment:
A bit of python [syntax](https://docs.python.org/3/library/constants.html#Ellipsis) I had not come across before.  In this context, a no-op equivalent to `pass`.